File: api/src/main/java/gustavo/ventieri/capitalmind/infrastructure/service/ExpenseService.java
package gustavo.ventieri.capitalmind.infrastructure.service;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import gustavo.ventieri.capitalmind.application.dto.expense.ExpenseRequestDto;
import gustavo.ventieri.capitalmind.application.dto.expense.ExpenseResponseDto;
import gustavo.ventieri.capitalmind.application.service.ExpenseServiceInterface;
import gustavo.ventieri.capitalmind.domain.expense.Expense;
import gustavo.ventieri.capitalmind.domain.user.User;
import gustavo.ventieri.capitalmind.infrastructure.exception.NotFoundException;
import gustavo.ventieri.capitalmind.infrastructure.mapper.expense.ExpenseMapper;
import gustavo.ventieri.capitalmind.infrastructure.persistence.ExpenseRepository;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ExpenseService implements ExpenseServiceInterface {

    private final ExpenseRepository expenseRepository; // Repositório de despesas
    private final UserService userService; // Serviço para validar e obter usuários
    private final ExpenseMapper expenseMapper; // Mapeador entre a entidade Expense e o DTO

    /**
     * Cria uma nova despesa associada a um usuário.
     */
    @Override
    public void create(ExpenseRequestDto expenseRequestDto) {
        // Valida e obtém o usuário
        User user = this.userService.validateAndGetUser(expenseRequestDto.userId());

        // Salva a despesa no banco de dados
        this.expenseRepository.save(
            new Expense(
                null,
                expenseRequestDto.name(),
                expenseRequestDto.description(),
                expenseRequestDto.category(),
                expenseRequestDto.price(),
                user,
                Instant.now(),
                Instant.now()
            )
        );
    }

    /**
     * Atualiza uma despesa existente.
     */
    @Override
    public void update(Long expenseId, ExpenseRequestDto expenseRequestDto) {
        // Busca a despesa no banco de dados
        Expense expense = this.expenseRepository.findById(expenseId)
            .orElseThrow(() -> new NotFoundException("Expense Not Found"));

        // Atualiza os campos da despesa
        expense.setName(expenseRequestDto.name());
        expense.setDescription(expenseRequestDto.description());
        expense.setCategory(expenseRequestDto.category());
        expense.setPrice(expenseRequestDto.price());

        // Salva a despesa atualizada no banco de dados
        this.expenseRepository.save(expense);
    }

    /**
     * Obtém todas as despesas associadas a um usuário.
     */
    @Override
    public List<ExpenseResponseDto> getAll(String userId) {
        // Valida e obtém o usuário
        User user = this.userService.validateAndGetUser(userId);

        // Recupera todas as despesas associadas ao usuário
        List<Expense> expenses = this.expenseRepository.findAllByUserData(user);

        // Mapeia as despesas para DTOs
        return expenses.stream()
            .map(expense -> expenseMapper.toDto(expense))
            .collect(Collectors.toList());
    }

    /**
     * Obtém uma despesa por ID.
     */
    @Override
    public ExpenseResponseDto getById(Long expenseId) {
        // Busca a despesa no banco de dados
        Expense expense = this.expenseRepository.findById(expenseId)
            .orElseThrow(() -> new NotFoundException("Expense Not Found"));

        // Mapeia a despesa para o DTO e retorna
        return expenseMapper.toDto(expense);
    }

    /**
     * Exclui uma despesa por ID.
     */
    @Override
    public void deleteById(Long expenseId) {
        // Verifica se a despesa existe
        if (!this.expenseRepository.existsById(expenseId)) {
            throw new NotFoundException("Expense Not Found");
        }

        // Exclui a despesa do banco de dados
        this.expenseRepository.deleteById(expenseId);
    }
}

File: src/shared/services/api/request/ExpenseService.ts
import { api } from "../config/axios/AxiosConfig";

export interface IExpense {
  expenseId: number;
  name: string;
  description: string;
  category: string;
  price: number;
  userId: string;
}

const getAll = async (id: string): Promise<IExpense[] | Error> => {
  try {
    const { data } = await api.get(`/expense/all/${id}`);

    if (data) {
      return data;
    }

    return [];
  } catch (error) {
    return new Error((error as { message: string }).message || "Error 404.");
  }
};

const create = async (
  dados: Omit<IExpense, "expenseId">
): Promise<string | Error> => {
  try {
    const { data } = await api.post<IExpense>("/expense/create", dados);

    if (typeof data === "string") return data;

    return new Error("Erro ao criar o registro.");
  } catch (error) {
    return new Error(
      (error as { message: string }).message || "Erro ao criar o registro."
    );
  }
};

const update = async (dados: IExpense): Promise<string | Error> => {
  try {
    const { data } = await api.put<IExpense>(
      `/expense/update/${dados.expenseId}`,
      dados
    );

    if (typeof data === "string") return data;

    return new Error("Erro ao criar o registro.");
  } catch (error) {
    return new Error(
      (error as { message: string }).message || "Erro ao criar o registro."
    );
  }
};

const deleteById = async (id: number): Promise<void | Error> => {
  try {
    await api.delete(`/expense/delete/${id}`);
  } catch (error) {
    return new Error(
      (error as { message: string }).message || "Erro ao apagar o registro."
    );
  }
};

export const ExpenseService = {
  getAll,
  create,
  deleteById,
  update,
};
